pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/CompoundDataBuffer.java

```diff
@@ -0,0 +1,665 @@
+ * under the License.
+ */
+package org.apache.pinot.segment.spi.memory;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.BufferUnderflowException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import org.roaringbitmap.buffer.ImmutableRoaringBitmap;
+
+
+/**
+ * A {@link DataBuffer} that is composed of multiple {@link DataBuffer}s that define a single contiguous buffer.
+ * <p>
+ * While reads and writes can span multiple buffers, there may be a performance impact when doing so.
+ * Therefore it is recommended to try to wrap independent buffers.
+ * <p>
+ * Once this class is built, buffers cannot be added or removed.
+ */
+public class CompoundDataBuffer implements DataBuffer {
+
+  private final DataBuffer[] _buffers;
+  private final long[] _bufferOffsets;
+  private int _lastBufferIndex = 0;
+  private final ByteOrder _order;
+  private final long _size;
+  private final boolean _owner;
+
+  /**
+   * Creates a compound buffer from the given buffers.
+   *
+   * @param buffers The buffers that will be concatenated to form the compound buffer.
+   * @param order The byte order of the buffer. Buffers in the array that have a different byte order will be converted.
+   * @param owner Whether this buffer owns the underlying buffers. If true, the underlying buffers will be released when
+   *              this buffer is closed.
+   */
+  public CompoundDataBuffer(DataBuffer[] buffers, ByteOrder order, boolean owner) {
+    _owner = owner;
+    _buffers = buffers;
+    _bufferOffsets = new long[buffers.length];
+    _order = order;
+    long offset = 0;
+    for (int i = 0; i < buffers.length; i++) {
+      if (buffers[i].size() == 0) {
+        throw new IllegalArgumentException("Buffer at index " + i + " is empty");
+      }
+      if (buffers[i].order() != _order) {
+        buffers[i] = buffers[i].view(0, buffers[i].size(), _order);
+      }
+    }
+    for (int i = 0; i < buffers.length; i++) {
+      _bufferOffsets[i] = offset;
+      long size = buffers[i].size();
+      offset += size;
+    }
+    _size = offset;
+  }
+
+  public CompoundDataBuffer(ByteBuffer[] buffers, ByteOrder order, boolean owner) {
+    this(Arrays.stream(buffers).map(PinotByteBuffer::wrap).toArray(DataBuffer[]::new), order, owner);
+  }
+
+  /**
+   * Creates a compound buffer from the given buffers.
+   * @param buffers The buffers that will be concatenated to form the compound buffer.
+   * @param order The byte order of the buffer. Buffers in the list that have a different byte order will be converted.
+   * @param owner Whether this buffer owns the underlying buffers. If true, the underlying buffers will be released when
+   *              this buffer is closed.
+   */
+  public CompoundDataBuffer(List<DataBuffer> buffers, ByteOrder order, boolean owner) {
+    this(buffers.toArray(DataBuffer[]::new), order, owner);
```

Comment:
Avoid functional programming APIs
```suggestion
    this(buffers.toArray(new DataBuffer[0]), order, owner);
```

Reply:
Done